elements/ip/ipfilter.hh

```diff
@@ -323,28 +372,55 @@ inline bool
 IPFilter::Primitive::negation_is_simple() const
 {
     if (_type == TYPE_PROTO)
-	return true;
+        return true;
     else if (_transp_proto >= 0)
-	return false;
+        return false;
     else
-	return _type == TYPE_HOST || (_type & TYPE_FIELD) || _type == TYPE_IPFRAG;
+        return _type == TYPE_HOST || (_type & TYPE_FIELD) || _type == TYPE_IPFRAG;
 }
 
 inline int
-IPFilter::match(const IPFilterProgram &zprog, const Packet *p)
+IPFilter::match(const IPFilterProgram &zprog, const Packet *p, const bool &caching)
 {
     int packet_length = p->network_length(),
-	network_header_length = p->network_header_length();
+    network_header_length = p->network_header_length();
     if (packet_length > network_header_length)
-	packet_length += offset_transp - network_header_length;
+        packet_length += offset_transp - network_header_length;
     else
-	packet_length += offset_net;
+        packet_length += offset_net;
 
-    if (zprog.output_everything() >= 0)
-	return zprog.output_everything();
-    else if (packet_length < (int) zprog.safe_length())
-	// common case never checks packet length
-	return length_checked_match(zprog, p, packet_length);
+    if (zprog.output_everything() >= 0) {
+        if (caching) {
+            (*_cache_misses_nb.findp(name()))++;
+        }
+        return zprog.output_everything();
+    }
+    else if (packet_length < (int) zprog.safe_length()) {
+        if (caching) {
+            (*_cache_misses_nb.findp(name()))++;
+        }
+        // common case never checks packet length
+        return length_checked_match(zprog, p, packet_length);
+    }
+
+    // Caching enabled
+    if (caching) {
+        if (_last_flow_id[name()]) {
```

Comment:
Er... multiple hashtable search in the fast path? :p Why not a per-element cache?

Reply:
Good catch, thanks! I was already fixing the problem as I realized myself how stupid it was :p

Reply:
Do not merge yet, as I want to test the performance of caching with my new testie.
I will give you the green flag once I finish.
Of course, feel free to propose more improvements ;)

Reply:
Performance looks good in general. With a small number of rules, there is a very light impact of caching on the latency, but with a large number of rules the performance gains can be substantial. Green flag from me :p